Src/Support/Google.Apis.Auth/OAuth2/ImpersonatedCredential.cs

```diff
@@ -0,0 +1,176 @@
+            /// <param name="options">The impersonation options.</param>
+            public Initializer(GoogleCredential sourceCredential, ImpersonationOptions options)
+                : base(String.Format(GoogleAuthConsts.IamAccessTokenEndpointFormatString, options.TargetPrincipal))
+            { 
+                SourceCredential = sourceCredential;
+                Options = options;
+            }
+
+            internal Initializer(ImpersonatedCredential other) : base(other.TokenServerUrl)
+            {
+                SourceCredential = other.SourceCredential;
+                Options = other.Options;
+            }
+        }
+
+        /// <summary>
+        /// Gets the source credential used as to acquire the impersonated credentials.
+        /// </summary>
+        public GoogleCredential SourceCredential { get; }
+
+        /// <summary>
+        /// Gets the impersonation options.
+        /// </summary>
+        public ImpersonationOptions Options { get; }
+
+        /// <summary>Constructs a new impersonated credential using the given initializer.</summary>
+        internal ImpersonatedCredential(Initializer initializer) : base(initializer)
+        {
+            SourceCredential = initializer.SourceCredential.ThrowIfNull("initializer.SourceCredential");
+            if (!(SourceCredential.UnderlyingCredential is UserCredential || SourceCredential.UnderlyingCredential is ServiceAccountCredential))
+            {
+                throw new InvalidOperationException("The underlying credential of source credential must be UserCredential or ServiceAccountCredential.");
+            }
+            SourceCredential = SourceCredential.CreateScoped(new string[] {GoogleAuthConsts.IamScope});
+            HttpClient.MessageHandler.Credential = SourceCredential.UnderlyingCredential as Http.IHttpExecuteInterceptor;
+            Options = initializer.Options;
+        }
+
+        /// <inheritdoc/>
+        IGoogleCredential IGoogleCredential.WithQuotaProject(string quotaProject) =>
+            new ImpersonatedCredential(new Initializer(this) { QuotaProject = quotaProject });
+
+        /// <inheritdoc/>
+        public override async Task<bool> RequestAccessTokenAsync(CancellationToken taskCancellationToken)
+        {
+            var bodyJson = new ImpersonationAccessTokenRequest
+            {
+                DelegateAccounts = Options.DelegateAccounts,
+                Scopes = Options.Scopes,
+                Lifetime = ((int)Options.Lifetime.TotalSeconds).ToString() + "s"
+            };
+            var body = NewtonsoftJsonSerializer.Instance.Serialize(bodyJson);
+            var content = new StringContent(body, Encoding.UTF8, "application/json");
+            
+            var response = await HttpClient.PostAsync(TokenServerUrl, content, taskCancellationToken).ConfigureAwait(false);
+            var newToken = await TokenResponse.FromHttpResponseAsync(response, Clock, Logger).ConfigureAwait(false);
```

Comment:
I am not sure if this is necessary. This basically means moving these methods to TokenRequestExtensions class. ParameterUtils won't be used because it url encodes the body instead of using json format.